### ISSUE TITLE
`append`/`concat` function

### ISSUE THREAD

**Issue by jariji** [JuliaLang/julia#53040]
I want a non-mutating version of `append!` that puts two sequences together.

```jl
julia> (append(x::T, y::T) where T<: AbstractVector) = append!(copy(x), y);

julia> (append(x::T, y::T) where T<:AbstractString)= join([x, y]);

julia> append([:a, :b], [:c, :d])
4-element Vector{Symbol}:
 :a
 :b
 :c
 :d

julia> append("ab", "cd")
"acdb"
```

It has

- `forall a b c, append(a, append(b, c)) == append(append(a, b), c)`
- `exists e, forall a, append(e, a) == append(a, e) == a`

When the argument types differ we could have some promotion logic.

**Comment by tecosaur** [comment]
What's wrong with `vcat`?

**Comment by jariji** [comment]
> What's wrong with `vcat`?


- `vcat` puts string arguments in a vector: `vcat("ab", "cd") == ["ab", "cd"]`, whereas I want `append("ab", "cd") == "abcd"`.
- `vcat` accepts scalars, which I don't want in a concatenation function: `vcat([:a], :b) == [:a, :b]`. https://github.com/JuliaLang/julia/issues/34395

**Comment by jakobnissen** [comment]
Makes sense. One more question: Should there be a fallback collection type, e.g. if you `append` two lazy iterators? It'd be more consistent with Base to return a `Vector` (or an `Array`?), but that behaviour implies it is a breaking change to override the default which I think is bad design.
Other options are:
* Do not implement a fallback
* Have the fallback simply call `Iterators.flatten`

**Comment by BioTurboNick** [comment]
`union` does most of what is requested, AFAICT, except that for string arguments it returns a vector of characters.

**Comment by barucden** [comment]
>union does most of what is requested

Does it? It returns unique elements:
```julia
julia> union([1,2,3], [1,2,3])
3-element Vector{Int64}:
 1
 2
 3
```
While `append` should return `[1,2, 3, 1, 2, 3]`.

**Comment by BioTurboNick** [comment]
LOL sorry, too early in the morning.

**Comment by mcabbott** [comment]
For precedent, maybe worth noting that `map(identity, x)` preserves Vector, String, Tuple, NamedTuple, gives an error on Dict & Set, and (perhaps unfortunately) makes an array from unknown iterators. `filter(Returns(true), x)` should preserve all of those, but is an error on unknown iterators.

Also that `stack` is in a sense another eager version of `Iterators.flatten`, but always makes an array: `stack(("abc", "def")) isa Matrix{Char}`. There was some discussion of whether we need another function which would be to `reduce(vcat, x)` as `stack` is to `reduce(hcat, x)`, but not of how it should behave on non-arrays, i.e. whether it should make a `Vector{Char}` or a `String`.

**Comment by jariji** [comment]
> There was some discussion of whether we need another function which would be to reduce(vcat, x) as stack is to reduce(hcat, x)

@mcabbott do you mean 

```jl
julia> reduce(vcat, [[1 0 0], [2 0 0]])
2×3 Matrix{Int64}:
 1  0  0
 2  0  0
```
? cuz that one adds a dimension; `append` doesn't add a dimension.

**Comment by jariji** [comment]
@mcabbott I'm not sure what `append(::Matrix, ::Matrix)` is supposed to do. Do you want it to be like `vcat` there?

**Comment by jariji** [comment]
@mcabbott Edited OP.

**Comment by jariji** [comment]
As for different-typed arguments, it could either error or promote. What do we think about, say, `append((1,), [2])`?

**Comment by gbaraldi** [comment]
https://clojuredocs.org/clojure.core/conj This looks to have some of the similar behaviour in clojure for reference
